File: src/shm_bridge_cpp/examples/example_usage.cpp
// example_usage.cpp — minimal demo of the shm_bridge C++ library.
//
// Shows the whole API with NO ROS: one Writer publishes a fake 4x4 RGB frame in a
// loop, one Reader attaches and reconstructs it. This is exactly what you'd embed
// in your own node — swap the fake buffer for real message data.
//
// Build: it's compiled as `example_usage` by this package's CMakeLists.
// Run two terminals:
//   ros2 run shm_bridge_cpp example_usage write
//   ros2 run shm_bridge_cpp example_usage read
#include <shm_bridge_cpp/shm_bridge.hpp>

#include <chrono>
#include <cstdio>
#include <cstring>
#include <string>
#include <thread>
#include <vector>

static int do_write() {
    // 4x4 RGB image = 48 bytes; size the buffer generously.
    shm_bridge::Writer w("example", 1920 * 1080 * 4);
    std::printf("[write] publishing 4x4 RGB to /dev/shm/example_* (Ctrl-C to stop)\n");

    std::vector<uint8_t> img(4 * 4 * 3);
    uint8_t tick = 0;
    while (true) {
        for (size_t i = 0; i < img.size(); ++i) img[i] = static_cast<uint8_t>(i + tick);
        w.write_flat(img.data(), img.size(),
                     /*width*/4, /*height*/4, /*channels*/3,
                     shm_bridge::DType::U8, "sensor_msgs/msg/Image", "/example");
        ++tick;
        std::this_thread::sleep_for(std::chrono::milliseconds(200));
    }
}

static int do_read() {
    shm_bridge::Reader r("example");      // throws if writer hasn't created it yet
    std::printf("[read] attached to /dev/shm/example_*\n");

    shm_bridge::Frame f;
    while (true) {
        if (r.read(f)) {
            std::printf("[read] seq=%u %s %ux%ux%u %zuB type=%s first3=[%u,%u,%u]\n",
                        f.seq,
                        f.encoding == shm_bridge::Encoding::FLAT ? "FLAT" : "CDR",
                        f.width, f.height, f.channels, f.data.size(),
                        f.type_name.c_str(),
                        f.data.size() > 2 ? f.data[0] : 0,
                        f.data.size() > 2 ? f.data[1] : 0,
                        f.data.size() > 2 ? f.data[2] : 0);
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(200));
    }
}

int main(int argc, char** argv) {
    std::string mode = argc > 1 ? argv[1] : "";
    if (mode == "write") return do_write();
    if (mode == "read") return do_read();
    std::fprintf(stderr, "usage: %s write|read\n", argv[0]);
    return 1;
}

File: src/shm_bridge_python/shm_bridge_python/example_usage.py
#!/usr/bin/env python3
"""example_usage.py — minimal demo of the shm_bridge_python library API.

No ROS needed. `write` publishes a fake 4x4 RGB frame in a loop; `read`
reconstructs it as a numpy array via the type-agnostic StreamReader. This is the
pattern you'd embed in your own code.

  ros2 run shm_bridge_python example_usage write
  ros2 run shm_bridge_python example_usage read

Or import directly in your own script:
  from shm_bridge_python import StreamReader, shm_contract as C
"""
import sys
import time

import numpy as np

from shm_bridge_python import shm_contract as C
from shm_bridge_python import StreamReader

NAME = "example_py"


def do_write():
    hp, fp, rp = C.shm_paths(NAME)
    _, hdr, _ = C.open_mmap(hp, C.HEADER_SIZE, create=True)
    _, frame, cap = C.open_mmap(fp, 1920 * 1080 * 4, create=True)
    C.write_recipe(rp, {
        "topic": "/example_py", "type_name": "sensor_msgs/msg/Image",
        "encoding": "flat", "dtype": "uint8", "shape": [4, 4, 3],
    })
    print(f"[write] publishing 4x4 RGB to /dev/shm/{NAME}_* (Ctrl-C to stop)")
    seq = 0
    tick = 0
    while True:
        img = ((np.arange(4 * 4 * 3) + tick) % 256).astype(np.uint8)
        payload = img.tobytes()
        seq += 1                                    # odd: writing
        hdr[0:4] = seq.to_bytes(4, "little")
        frame[0:len(payload)] = payload
        hdr[0:C.HEADER_SIZE] = C.pack_header(
            seq, C.ENC_FLAT, len(payload), 4, 4, 3,
            C.ID_BY_DTYPE["uint8"], time.time_ns(), "sensor_msgs/msg/Image")
        seq += 1                                    # even: stable
        hdr[0:4] = seq.to_bytes(4, "little")
        tick += 1
        time.sleep(0.2)


def do_read():
    rd = StreamReader(NAME)                          # type-agnostic
    print(f"[read] attached to /dev/shm/{NAME}_*")
    while True:
        obj, hdr = rd.read_frame()
        if obj is not None:
            if isinstance(obj, np.ndarray):
                print(f"[read] seq={hdr['seq']} FLAT ndarray "
                      f"shape={obj.shape} dtype={obj.dtype} first3={obj.flat[:3]}",
                      flush=True)
            else:
                print(f"[read] seq={hdr['seq']} CDR {type(obj).__name__}", flush=True)
        time.sleep(0.2)


def main():
    mode = sys.argv[1] if len(sys.argv) > 1 else ""
    if mode == "write":
        do_write()
    elif mode == "read":
        do_read()
    else:
        print("usage: example_usage write|read")
        return 1


if __name__ == "__main__":
    raise SystemExit(main())
